relay-filter/src/generic.rs

```diff
@@ -27,44 +26,178 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+        project_filters,
+        global_filters,
+        #[cfg(test)]
+        VERSION,
+    );
 
-    for filter_config in config.filters.iter() {
-        if !filter_config.is_empty() && matches(event, filter_config.condition.as_ref()) {
-            return Err(FilterStatKey::GenericFilter(filter_config.id.clone()));
+    for filter_config in filters {
+        if filter_config.is_enabled && matches(event, filter_config.condition) {
+            return Err(FilterStatKey::GenericFilter(filter_config.id.to_owned()));
         }
     }
 
     Ok(())
 }
 
+/// Returns an iterator that yields merged generic configs.
+///
+/// Since filters of project and global configs are complementary and don't
+/// provide much value in isolation, both versions must match [`VERSION`] to be
+/// compatible.  If filters aren't compatible, an empty iterator is returned.
+///
+/// If filters are compatible, an iterator over all filters is returned. This
+/// iterator yields filters according to the principles below:
+/// - Filters from project configs are evaluated before filters from global
+/// configs.
+/// - No duplicates: once a filter is evaluated (yielded or skipped), no filter
+/// with the same id is evaluated again.
+/// - If a filter with the same id exists in projects and global configs, both
+/// are merged and the filter is yielded. Values from the filter in the project
+/// config are prioritized.
+fn merge_generic_filters<'a>(
+    project: &'a GenericFiltersConfig,
+    global: Option<&'a GenericFiltersConfig>,
+    #[cfg(test)] version: u16,
+) -> impl Iterator<Item = MergedFilter<'a>> {
+    #[cfg(not(test))]
+    let version = VERSION;
+
+    let is_supported =
+        project.version <= version && global.map_or(true, |gf| gf.version <= version);
+
+    is_supported
+        .then(|| {
+            DynamicGenericFiltersConfigIter::new(&project.filters, global.map(|gc| &gc.filters))
+        })
+        .into_iter()
+        .flatten()
+}
+
+/// Iterator over the generic filters of the project and global configs.
+struct DynamicGenericFiltersConfigIter<'a> {
+    /// Generic project filters.
+    project: &'a IndexMap<String, GenericFilterConfig>,
+    /// Index of the next filter in project configs to evaluate.
+    project_index: usize,
+    /// Generic global filters.
+    global: Option<&'a IndexMap<String, GenericFilterConfig>>,
+    /// Index of the next filter in global configs to evaluate.
+    global_index: usize,
+}
+
+impl<'a> DynamicGenericFiltersConfigIter<'a> {
+    pub fn new(
+        project: &'a IndexMap<String, GenericFilterConfig>,
+        global: Option<&'a IndexMap<String, GenericFilterConfig>>,
+    ) -> Self {
+        DynamicGenericFiltersConfigIter {
+            project,
+            project_index: 0,
+            global,
+            global_index: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for DynamicGenericFiltersConfigIter<'a> {
+    type Item = MergedFilter<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
Creating two iterators before `impl Iterator` makes the iterator implementation easier to understand, but adds unneeded complexity on typing these iterators in the struct: exact type of the iterator including the closure of the filtering that must be known at compile time and creating yet another struct is not worth it.